### DISCUSSION TITLE
Monitoring: Exporter for Prometheus

### DISCUSSION THREAD

**Discussion by mario-zelger** [ravendb/ravendb#14325] [General]
Hi guys

Are you by any chance planning on providing an official [exporter for Prometheus](https://prometheus.io/docs/instrumenting/exporters/) as you did with the [Telegraf Plugin](https://ravendb.net/docs/article-page/5.3/http/server/administration/monitoring) to easily scrape RavenDB metrics inside a Kubernetes Cluster?

Cheers,
Mario

**Comment by ml054** [comment]
Hi @mario-zelger 

Direct prometheus exporter is on our roadmap. You can subscribe to this issue https://issues.hibernatingrhinos.com/issue/RavenDB-17674 to track progress. 

For now, feel free to use telegraf plugin to feed prometheus with metrics. Basic instructions to set this up are here: https://issues.hibernatingrhinos.com/issue/RavenDB-17674#focus=Comments-67-274223.0-0

**Comment by sudiplamsal00** [comment]
The RavenDB endpoints provides iops and throughput data in <your server URL>/admin/monitoring/v1/server. However, that metric is not available to be scrap. Is there any plan to upgrade for IoWriteOperations,IoReadOperations,ReadThroughputInKb,WriteThroughputInKb and QueueLength. Also the  other metrics in json data, so expecting update of endpoint fields.

**Comment by ml054** [comment]
What do you mean by 'is not avaialbe to be scrap'? By whom?

**Comment by sudiplamsal00** [comment]
As per this doc https://ravendb.net/docs/article-page/5.4/csharp/server/administration/monitoring, I don't see field name for iops and throughput. Also, i am currently using telegraf, influxdb & grafana for monitoring and we need charts for iops and throughput, which seems to be not available.

**Reply by ayende** [reply]
They are there, see: `IoWriteOperations` and `IoReadOperations` here:
http://live-test.ravendb.net/admin/monitoring/v1/databases

**Reply by sergiomcalzada** [reply]
Yep, they are there, but the telegraf pluging don't gather them

https://github.com/influxdata/telegraf/blob/master/plugins/inputs/ravendb/ravendb.go#L214

**Reply by ml054** [reply]
@sergiomcalzada Here is ticket, so you can track progress:
https://issues.hibernatingrhinos.com/issue/RavenDB-18715/Expose-the-new-disk-metrics-fields-in-Telegraf

**Reply by ml054** [reply]
So basically we need to update telegraf plugin and wait for telegraf release.

**Comment by ml054** [comment]
Please notice that those metrics are right now exposed in monotoring endpoints only for Linux machines. 

For windows we have open ticket: https://issues.hibernatingrhinos.com/issue/RavenDB-18979/Track-disk-metrics-on-Windows

**Comment by mkocaks** [comment]
Dear Team,
We are testing telegraf with Prometheus client and ravendb, and looking at replacing our old ravenexporter alerts with telegraf, however one of metrics we use from raven exporter namely ravendb_is_leader cannot be replaced to make use of the telegraf metrics...
The Ravendb_is_leader metric is either a 1 or a 0 however nothing like this exists for telegraf, we tried looking at ravendb_server_cluster_current_term however this is a constantly ever increasing metric and always returns a result regardless due to the way this metric has been implemented.
Is there any possibility to add metric to show when a leader change occurs in raven ? (Pretty please! :))

**Reply by ml054** [reply]
@mkocaks Thanks for letting us know. We will try to take your suggestion into account. I updated the ticket: https://issues.hibernatingrhinos.com/issue/RavenDB-18715/Expose-the-new-disk-metrics-fields-in-Telegraf

**Reply by mkocaks** [reply]
Done! I guess no real eta for this ?

**Reply by mkocaks** [reply]
https://issues.hibernatingrhinos.com/issue/RavenDB-20649/Ravendb-leader-change-metric-needed

**Reply by ml054** [reply]
We have plan to cover that in one of upcoming sprints - then we will have to wait for telegraf to merge/release.

**Reply by mkocaks** [reply]
Thank you!, just a heads up the "leader" value should be the metric and not a label value in the metric, for example:

It isn't recommended to put frequently changed values into labels, because this may result in [high cardinality issues](https://docs.victoriametrics.com/FAQ.html#what-is-high-cardinality). Ideally all the labels must be static. So the usage label must be put into time series value instead. E.g. the resulting metric should look like usage{mountpoint="test"} 90 when expressed in [Prometheus text exposition format](https://github.com/Showmax/prometheus-docs/blob/master/content/docs/instrumenting/exposition_formats.md#text-format-details). In this case the PromQL query, which returns disk space usage greater than 80%, will look as simple as usage > 80.

**Reply by Lwiel** [reply]
We expose `cluster_node_state` metric that takes values `Passive`, `Candidate`, `Follower`, `LeaderElect`, `Leader`. You can use [Telegraf processor plugins](https://docs.influxdata.com/telegraf/v1.26/plugins/#processor-plugins) to transform data the way it fits your needs.

**Reply by mkocaks** [reply]
These ones
![image](https://github.com/ravendb/ravendb/assets/87523722/1cefa0aa-1ba2-4497-873d-d74d76172fd4)

**Reply by Lwiel** [reply]
These changes haven't been yet included in the newest Telegraf release (v1.27.1), so you need to wait until they include them, or build from source. They're also only exposed on Linux.

**Reply by mkocaks** [reply]
Thank you, we have noticed that the changes is only n the master branch - sorry for that, going back to the "isleader" metric, the processer pluigns I have looked at (enum and parser) does not do what we need...
We need an actual metric with a value of 1 or 0 (true or false) that shows us who is the leader....enum only creates a new tag/field (or overwrites) and parser simply not working at all....any ideas ?
We use alertmanager to look at the last 5 mins of any leader chnages and these promql's only use values of the metric not tags values...

**Reply by mkocaks** [reply]
This is what I tried amongst  other things thinking we will get a brand new metric with a value of 0 or 1, or 2, or 3, or 4...

[agent]
      collection_jitter = "0s"
      debug = false
      flush_interval = "10s"
      flush_jitter = "0s"
      hostname = "$HOSTNAME"
      interval = "10s"
      logfile = ""
      metric_batch_size = 1000
      metric_buffer_limit = 10000
      omit_hostname = false
      precision = ""
      quiet = false
      round_interval = true
    [[processors.enum]]
       [[processors.enum.mapping]]
        dest = "cluster_node_role"
        field = "cluster_node_state"
        tag = "cluster_node_state"
        [processors.enum.mapping.value_mappings]
            Candidate = "1"
            Follower = "2"
            Leader = "4"
            LeaderElect = "3"
            Passive = "0"
    [[processors.parser]]
      data_format = "prometheus"
      drop_original = false
      merge = "override"
      parse_tags = [
        "cluster_node_role"
      ]


    [[outputs.prometheus_client]]
      listen = ":9273"
      string_as_label = true

    [[inputs.ravendb]]
      url = "http://ravendb-0.ravendb:8080"
    [[inputs.ravendb]]
      url = "http://ravendb-1.ravendb:8080"
    [[inputs.ravendb]]
      url = "http://ravendb-2.ravendb:8080"

    [[inputs.internal]]
      collect_memstats = false

**Reply by ml054** [reply]
btw. This was merged today - https://github.com/ravendb/ravendb/pull/16712

It brings direct support for promethues. 
For each node we expose cluster state as number. 
Example:
```
# HELP ravendb_cluster_node_state Current node state, Values: 0 = Passive, 1 = Candidate, 2 = Follower, 3 = LeaderElect, 4 = Leader
# TYPE ravendb_cluster_node_state gauge
ravendb_cluster_node_state 4
```

**Reply by mkocaks** [reply]
Wow, pretty cool for the direct integration !! We just need to plan our next RavenDB Upgrade now :-D

Thanks yep and I managed to add these values as tags to existing metrics via enum processor plugin but wont help...
Also this means we will need to upgrade raven :-(
If the parser side of it worked then this would of done the trick for us... but something is not right with the parser config and or source code

**Reply by mkocaks** [reply]
Maybe because the Prometheus client already uses its own parser and ignores anything under processors
 maybe ?

**Reply by ml054** [reply]
@mkocaks I'm not sure I follow.

**Reply by mkocaks** [reply]
So for example the raven exporter metric "ravendb_is_leader" value returns a 0 (for follower) or a 1 (for leader) - we monitor the rate at which this value changes in a time period and alert when a change occurs.
As far as I know there is no way to do such a thing for a value of a tag for a metric, thus we need a dedicated metric with a value of 0 or 1 for when a leader change occurs (i.e. node value for this metric will change to 1)
Each raven node will have this metric (0 or 1) in the cluster....

**Reply by mkocaks** [reply]
I was being an idiot!
Using enum plugin for telegraf and re mapping the cluster_node_state string values to an integer value then using the string as label = false under the outputs settings of the telegraf config map done the trick! Thanks!

**Reply by mkocaks** [reply]
This creates a new metric in Prometheus with the values needed

**Comment by SbiCA** [comment]
@ml054  I guess it belongs in here but I think it would be good to have sort of a bootstrap story including altering, dashboards similar to what's already there. There seem to be quite a lot of issues raven already has alerts in the studio but probably could be mapped to alerts that can be triggered (of course assuming the metrics are exposed)

https://grafana.com/oss/prometheus/exporters/postgres-exporter/

**Reply by ml054** [reply]
I'm not sure I follow your comment.

**Reply by SbiCA** [reply]
I mean have already a compiled list of reasonable alerts as they do with postgres e.g. see https://grafana.com/oss/prometheus/exporters/postgres-exporter/?tab=alerting-rules
```
 - alert: PostgreSQLHighConnections
      expr: sum(pg_stat_activity_count) by (instance) > (sum(pg_settings_max_connections) by (instance) - sum(pg_settings_superuser_reserved_connections) by (instance)) * 0.8
      for: 10m
      labels:
        severity: email
      annotations:
        summary: "{{ $labels.instance }} is over 80% of max Postgres connections."
        description: "{{ $labels.instance }} is exceeding 80% of the currently configured maximum Postgres connection limit (current value: {{ $value }}s). Please check utilization graphs and confirm if this is normal service growth, abuse or an otherwise temporary condition or if new resources need to be provisioned (or the limits increased, which is mostly likely)."

```

I assume you already have that as part of raven studio since you'd usually get things like `SlowDisk` or `Query not using paging` and I guess there could be a reasonable default list what could be alerted on if you're running raven
